nrf91m1: Split signing pipeline with isolated access to Vault signing (#372)
nrfconnect/ncs-serial-modem: doc/app/nrf91m1/signing.rst

.. _nrf91m1_signing:

nRF91M1 signing procedures
###########################

All commands below assume ``ncs-serial-modem/app`` as the working directory.

.. note::

   **Serial Modem connection**

   After flashing, a working Serial Modem requires the correct UART wiring — in
   particular the **DTR** pin wired to **GND** when using a PC host. See the
   :ref:`uart_configuration` documentation for pin mapping, signal descriptions,
   and host setup.

Signing pipeline
================

``sign-hashes.sh`` is the only step that runs in the secure environment and the
only script that contacts Vault. All other scripts run on a normal build machine
and never touch private key material.

Typical split:

1. **Build machine** — ``build-unsigned.sh`` produces ``manifest-tosign.env``
   (opaque base64 hashes, no secrets).
2. **Secure environment** — carries in ``manifest-tosign.env`` and ``sign-hashes.sh``,

>    authenticates to Vault, produces ``manifest-signed.env`` (signatures, no key material).

Author: > All commands below assume ``ncs-serial-modem/app`` as the working directory.

The script exists.

>    authenticates to Vault, produces ``manifest-signed.env`` (signatures, no key material).

Author: This was actually a correct observation. The build-unsigned.sh with the .gitignore `build-*`. Renamed and added.

.. _nrf91m1_signing:

nRF91M1 signing procedures
###########################

All commands below assume ``ncs-serial-modem/app`` as the working directory.

.. note::

   **Serial Modem connection**

   After flashing, a working Serial Modem requires the correct UART wiring — in
   particular the **DTR** pin wired to **GND** when using a PC host. See the
   :ref:`uart_configuration` documentation for pin mapping, signal descriptions,

>    and host setup.

Reviewer: Shouldn't this be called signaling rather than wiring?
Does the protocol require that once the transfer completes then the DTR is pulled low? What is expected value of the DTR during the transfer, should it be left floating or held high? What is the high value for the DTR, Vcc?

Signing pipeline
================

All build-side scripts are Python and run on a normal build machine, never
touching private key material. For the secure signing step, both
``sign_hashes.py`` (requires Python ≥ 3.9) and ``sign-hashes.sh`` (requires
only bash and the vault CLI) are provided — use the shell version on signing

> machines without Python.

Reviewer: ```suggestion
All build-side scripts are Python and are supposed to be run on a ordinary build machine, with **no access to a private key material**. For the secure signing step ``sign_hashes.py``  and ``sign-hashes.sh``; bash version is provided for signing on machines with no python. 
``sign-hashes.sh`` requires python version higher than 3.9.
``sign_hashes.sh`` has no external requirements except bash at version x/yz/
```
The bash version should also be provided, you have tested the script on bash at some version, that version would be the minimal one.
Also, bash script uses '[' which is alias for external tool `test`.

Typical split:

1. **Build machine** — ``sign_build_unsigned.py`` produces ``manifest-tosign.env``
   (opaque base64 hashes, no secrets).
2. **Secure environment** — carries in ``manifest-tosign.env`` and
   ``sign_hashes.py`` (or ``sign-hashes.sh``), authenticates to Vault, produces
   ``manifest-signed.env`` (signatures, no key material).
3. **Build machine** — ``sign_assemble.py`` applies the signatures and produces
   the flashable images.

To authenticate in the secure environment:

.. code-block:: sh

   export VAULT_ADDR=https://vault.example.com
   export VAULT_TRANSIT_MOUNT=vault/location
   vault login

When a procedure involves two ``sign_hashes.py`` calls (MCUboot update), the
Vault session must still be valid for the second call, or re-authenticate.

.. warning::

>    **Build machine trust boundary**

Reviewer: ```suggestion
   **Build machine's trust boundaries**
```

or

```suggestion
   **Trust boundaries of a build machine**
```

``--override-mcuboot-version <N>`` overrides this value (debugging only).

.. code-block:: sh

   ./nrf91m1/scripts/sign_build_unsigned.py --dev

   # In secure environment:
   ./nrf91m1/scripts/sign_hashes.py \
       --in  ./signing-out/unsigned/manifest-tosign.env \
       --out ./signing-out/manifest-signed.env

   ./nrf91m1/scripts/sign_assemble.py \
       --signed       ./signing-out/manifest-signed.env \
       --unsigned-dir ./signing-out/unsigned

   ./nrf91m1/scripts/sign_prepare_mcuboot.py

   # Another round in secure environment:
   # Re-authenticate to Vault if the session has expired
   ./nrf91m1/scripts/sign_hashes.py \
       --in  ./signing-out/manifest-mcuboot-tosign.env \
       --out ./signing-out/manifest-mcuboot-signed.env

   ./nrf91m1/scripts/sign_assemble_mcuboot.py \
       --mcuboot-signed ./signing-out/manifest-mcuboot-signed.env

   ./scripts/sm_dfu_host.py --port /dev/ttyACM0 --baudrate 115200 \
       --type mcuboot-bootloader \
       --file ./signing-out/release/signed_by_mcuboot_and_b0_mcuboot_s1_variant.bin

   nrfutil device reset

Expected: MCUboot version counter using ``AT#XBOOTINFO=0`` returns 2.

B0 key rotation

> ===============

Reviewer 2: ```suggestion
B0 key revocation
=================
```

B0 provisions all public keys found in ``nrf91m1/certs/{env}/b0/`` at initial
flash. The active signing key advances monotonically via the hardware counter;

> once a key is rotated out, it is permanently revoked.

Reviewer 2: ```suggestion
earlier keys are permanently revoked if a later key passes verification.
```
